ceph-volume: terminal: encode unicode when writing to stdout

tchaikov opened the pull request:
python determins the encoding of stdout and stderr based on the LC_CTYPE
and PYTHONIOENCODING env variable, by default, python3's sys.stdout uses
'utf-8' as its encoding, so it will be able to write unicode string even
the stdout is not attached to a tty device. but when it comes to
python2, it will default to ascii if neither of these variabls is set.
so, if we are writing unicode using `_Write` in an environment where
LC_CTYPE and/or PYTHONIOENCODING are using non UTF-8 encoding, it chokes
by raising `UnicodeEncodeError` exception.

in this change, we add a wrapper around `_Write._writer` so it is able
to write unicode string in such a non-unicode-friendly environment.

for more info related the encoding of stdout and stderr, see
https://docs.python.org/3/using/cmdline.html#envvar-PYTHONIOENCODING .

Signed-off-by: Kefu Chai <kchai@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

tchaikov commented:
jenkins test ceph-volume lvm all

tchaikov commented:
jenkins test ceph-volume tox

tchaikov commented:
retest this please.

alfredodeza commented:
This looks good I think, but would love to see some unit tests that force these conditions to be handled. The class already allows for inputs and fake/stub std handlers. Would you mind adding a few tests?

tchaikov commented:
@alfredodeza what do you mean by "std handlers"?

alfredodeza commented:
I mean that you can inject the object that writes (defined as the `_writer` argument). I am surprised I couldn't find any tests for this already :(

tchaikov commented:
@alfredodeza i added a mock instead. i think `_Write` is not part of the interface exposed by `terminal`, so we should not test this module using `_Write`. could you please take another look?

alfredodeza commented on src/ceph-volume/ceph_volume/tests/test_terminal.py:
```diff
@@ -1,4 +1,10 @@
+
+import codecs
+import io
+import mock
```
this is not going to work :( ceph-volume (and ceph) do not depend on Mock, and although it is being added to tox.ini as a dependency it will break package workflows that require declaring dependencies for running tests.

Instead, we can use py.test's own mock framework (available as a test fixture) called `monkeypatch`

tchaikov replied:
thanks. TIL.

alfredodeza commented on src/ceph-volume/ceph_volume/tests/test_terminal.py:
```diff
@@ -66,3 +72,44 @@ def test_subcommand_found_and_dispatched_with_errors(self):
+
+class TestWriteUnicode(object):
+
+    octpus_and_squid_en = u'octpus and squid'
```
These should be part of a `setup()` method

alfredodeza commented on src/ceph-volume/ceph_volume/tests/test_terminal.py:
```diff
@@ -66,3 +72,44 @@ def test_subcommand_found_and_dispatched_with_errors(self):
+        with mock.patch('sys.stdout', new=stream):
+            terminal.stdout(self.message)
+            sys.stdout.flush()
+            assert self.octpus_and_squid_en.encode(encoding) in buffer.getvalue()
```
making this a fixture, and allowing buffer to be part of setup along with the message, this would become:


    @pytest.fixture
    def stream(buffer):
        # mock a stdout with given encoding
        if sys.version_info >= (3, 0):
            stdout = sys.stdout
            stream = io.TextIOWrapper(buffer,
                                      encoding=encoding,
                                      errors=stdout.errors,
                                      newline=stdout.newlines,
                                      line_buffering=stdout.line_buffering)
        else:
            stream = codecs.getwriter(encoding)(buffer)
            # StreamWriter does not have encoding attached to it, it will ask
            # the inner buffer for "encoding" attribute in this case
            stream.encoding = encoding
        return stream

    class TestWriteUnicode(object):

        def setup(self):
            octpus_and_squid_en = u'octpus and squid'
            octpus_and_squid_zh = u'章鱼和鱿鱼'
            self.message = octpus_and_squid_en + octpus_and_squid_zh
            self.buffer = io.BytesIO()

        def test_utf8_writer(self, stream, monkeypatch):
            stream = stream(self.buffer)
            monkeypatch.stattr('sys.stdout', stream)
            terminal.stdout(self.message)
            sys.stdout.flush()
            assert self.octpus_and_squid_en.encode('utf8') in self.buffer.getvalue()

tchaikov commented:
@alfredodeza i incorporated your change into this PR.

changelog

- extracted the fixture into `stream` using `pytest.fixture`.
- prepare the unicode string and the `encode("replace")` string using `setup()`

tchaikov commented on src/ceph-volume/ceph_volume/tests/test_terminal.py:
```diff
@@ -66,3 +71,51 @@ def test_subcommand_found_and_dispatched_with_errors(self):
+        self._test_writer('utf8', stream, monkeypatch)
+
+    def _test_writer(self, encoding, stream, monkeypatch):
+        buffer = io.BytesIO()
```
buffer is not used by all tests so i leave it in this function.

alfredodeza replied:
you took what I suggested and made it a thousand times better, this is great work here @tchaikov

tchaikov commented:
@alfredodeza the tests pass locally. mind taking a look?

alfredodeza commented:
jenkins test ceph-volume tox

alfredodeza commented:
py.test is getting into trouble because the test is patching `sys` globally, vs. where it is being used:

    E       AttributeError: '_io.FileIO' object has no attribute 'newlines'

I think the patching should be done on the module, e.g.:

    monkeypatch.setattr(ceph_volume.terminal.sys, 'stdout', stream(buffer, encoding))

tchaikov commented:
jenkins test ceph-volume tox

tchaikov commented:
@alfredodeza i think the root cause is that pytest captures `sys.stdout` using its own `pytest.capture.EncodedFile` which does not have the attributes .

i updated the test to disable the `capsys`. could you take another look ?

tchaikov commented on src/ceph-volume/ceph_volume/terminal.py:
```diff
@@ -80,11 +81,33 @@ def make(cls, string):
+            _writer = sys.stdout
+        self._writer = _Write._unicode_output_stream(_writer)
+        if _writer is sys.stdout:
+            sys.stdout = self._writer
```
@alfredodeza i realized that i should restore the `sys.stdout` if i am moving away its buffer from it, which practically renders `sys.stdout` closed.

tchaikov commented on src/ceph-volume/ceph_volume/tests/test_terminal.py:
```diff
@@ -66,3 +71,50 @@ def test_subcommand_found_and_dispatched_with_errors(self):
+        with capsys.disabled():
+            # we want to have access to the sys.stdout's attributes in
+            # make_stream(), not the ones of pytest.capture.EncodedFile
+            writer = stream(buffer, encoding)
```
also, instead of passing a temporary instance of writer. i found that i need to keep it alive during the test.

tchaikov commented:
jenkins test ceph-volume tox

tchaikov commented:
retest this please.

tchaikov commented:
@iotcg could you please file a bug if you run into any problem with the symptom and reproduce steps. i cannot tell what "the problem" you are facing without reading your mind.

iotcg commented:
@tchaikov 
Please check https://tracker.ceph.com/issues/41609
